File: spark/python_code/ingestionframework-master/TestData/VenkatBijjam/read_file.py
# -*- coding: utf-8 -*-
"""
Created on 04/12/2020 2:01 PM

@author : Nirav Langaliya
"""
from dependencies.schema_utilties import get_file_reader, get_options

def file_reader(
        path,
        data_format,
        schema_file,
        num_partittions,
        selectCol,
        filter
):
    options = get_options(
        data_format.header,
        data_format.delimiter,
        data_format.inferSchema,
        data_format.multiLine,
        data_format.roottag,
        data_format.rowtag
    )
    reader = get_file_reader(
        schema_file,
        options
    )

    file_format = str(data_format.dataFormat).lower()

    if file_format == 'csv':
        if filter.lower() == "none":
            dataframe = reader.csv(path).select(selectCol)
        else:
            dataframe = reader.csv(path).select(selectCol).filter(filter)
    elif file_format == "json":
        if filter.lower() == "none":
            dataframe = reader.json(path).select(selectCol)
        else:
            dataframe = reader.json(path).select(selectCol).filter(filter)
    elif file_format == "parquet":
        if filter.lower() == "none":
            dataframe = reader.parquet(path).select(selectCol)
        else:
            dataframe = reader.parquet(path).select(selectCol).filter(filter)
    elif file_format == "delta":
        if filter.lower() == "none":
            dataframe = reader.format("delta").load(path).select(selectCol)
        else:
            dataframe = reader.format("delta").load(path).select(selectCol).filter(filter)
    elif file_format == "xml":
        if filter.lower() == "none":
            dataframe = reader.format("com.databricks.spark.xml").load(path).select(selectCol)
        else:
            dataframe = reader.format("com.databricks.spark.xml").load(path).select(selectCol).filter(filter)
    elif file_format == "fixed":
        if filter.lower() == "none":
            dataframe = read_fixed_file(path, data_format, num_partittions).select(selectCol)
        else:
            dataframe = read_fixed_file(path, data_format, num_partittions).select(selectCol).filter(filter)
    elif file_format == 'avro' :
        """
        need to develop funcationality
        """
        dataframe = reader.format("avro").load(path)
        pass
    else:
        raise Exception(" Unknown data format! ")

    if num_partittions > 0:
        dataframe = dataframe.repartition(num_partittions)

    return dataframe


def read_fixed_file(
        path,
        data_format,
        num_partitions
):
    from dependencies.spark import spark_session
    from dependencies.schema_utilties import data_type_dict
    from pyspark.sql import functions as F
    from pyspark.sql.types import StringType
    import numpy

    if data_format.colNameTypeLength is None:
        raise Exception(" Missing fixed-length file parameters")

    column_length = []
    column_name = []
    column_datatype = []

    for col in data_format.colNameTypeLength:
        column_length.append(col.columnLength)
        column_datatype.append(col.columnType)
        column_name.append(col.columnName)

    cumulative_length = numpy.cumsum([1] + column_length)

    dataframe = spark_session.read.text(path).withColumnRenamed("value", "stringColumn")

    if data_format.header:
        initial_record = str(dataframe.first()[0])
        dataframe = dataframe.filter(F.col("stringColumn") != initial_record)

    if num_partitions > 0:
        dataframe = dataframe.repartition(num_partitions)

    for columns in zip(column_name, cumulative_length.tolist(), column_length):
        dataframe = dataframe.withColumn(
            columns[0],
            F.trim(
                F.substring(
                    F.col("stringColumn"),
                    columns[1],
                    columns[2]
                )
            )
        )

    dataframe = dataframe.drop(F.col("stringColumn"))

    for columns in zip(column_name, column_datatype):
        if columns[1] != "str":
            dataframe = dataframe.withColumn(
                columns[0],
                F.col(columns[0]).cast(data_type_dict.get(columns[1], StringType()))

            )

    return dataframe

File: spark/javaingestionframework-master/src/main/java/com/nisum/dataingestionframework/config/ReadFile.java
package com.nisum.dataingestionframework.config;

import com.nisum.dataingestionframework.dependencies.DateTimeUtilities;
import com.nisum.dataingestionframework.dependencies.SchemaUtilities;
import com.nisum.dataingestionframework.dependencies.ZipUnzipUtility;
import com.nisum.dataingestionframework.exception.ResourceNotFoundException;
import com.nisum.dataingestionframework.model.ColumnNameTypeLength;
import com.nisum.dataingestionframework.model.DatasetModel;
import com.nisum.dataingestionframework.model.FileInput;
import com.nisum.dataingestionframework.util.Constants;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.sql.*;
import org.apache.spark.sql.types.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.util.*;

/* @author Hari Chandra Prasad Nimmagadda */

public class ReadFile implements Serializable {

    private static final Logger logger = LoggerFactory.getLogger(ReadFile.class);

    Dataset<Row> dataFrame;
    Dataset<Row> newdf;
    SchemaUtilities schemaUtilities = new SchemaUtilities();
    DatasetModel datasetModel = new DatasetModel();
    Map<String, Dataset<Row>> dataSetMap = new HashMap<String, Dataset<Row>>();
    //StructType schema = new StructType();
    SparkSession sparkSession = SparkSession.builder().master("local[*]").appName("DataIngestionApp2").getOrCreate();


    ZipUnzipUtility zipUnzipUtility = new ZipUnzipUtility();
    DateTimeUtilities dateTimeUtilities = new DateTimeUtilities();

   public Map<String, Dataset<Row>> fileReader(String dataFrameName, FileInput fileInput) {

        boolean header = fileInput.getDataFormat().isHeader();
        String delimiter = fileInput.getDataFormat().getDelimiter();
        boolean inferschema = fileInput.getDataFormat().isInferSchema();
        boolean multiline = fileInput.getDataFormat().isMultiLine();
        String roottag = fileInput.getDataFormat().getRoottag();
        String rowtag = fileInput.getDataFormat().getRowtag();
        Optional<String> archivalPath = Optional.ofNullable(fileInput.getArchivalPath());

       String fileFormat = fileInput.getDataFormat().getDataFormat().toLowerCase();
       String archivePath = fileInput.getArchivalPath();

       Map options =  schemaUtilities.getOptions(header, delimiter, inferschema, multiline, roottag, rowtag);

       DataFrameReader reader = schemaUtilities.getFileReader(fileInput.getEnforceSchemaFile(), options);

        String path = fileInput.getFileInPath();

       if (fileInput.getNumPartitions() > 0)
           dataFrame = dataFrame.repartition(fileInput.getNumPartitions());

       fileInput.getSelectCol().stream()
                        .map(s -> dataFrame = dataFrame.select(s)
                            );

      archivalPath.ifPresent(s ->
              {
                  try {
                      zipUnzipUtility.unzip(s, archivePath);
                  } catch (Exception e) {
                      e.printStackTrace();
                  }
              }
              );

        logger.info("Before Execution of Read File " + dateTimeUtilities.getCurrentDateTime());

       if (fileFormat.equalsIgnoreCase(Constants.CSV))
         // dataFrame = reader.schema(schemaUtilities.getCsvSchema()).csv(path);
           dataFrame = reader.csv(path);
        else if(fileFormat.equalsIgnoreCase(Constants.JSON))
            dataFrame = reader.json(path);
        else if(fileFormat.equalsIgnoreCase(Constants.PARQUET))
            dataFrame = reader.parquet(path);
        else if(fileFormat.equalsIgnoreCase(Constants.DELTA))
            dataFrame = reader.format(Constants.DELTA).load(path);
        else if (fileFormat.equalsIgnoreCase(Constants.XML))
           dataFrame = reader.format("com.databricks.spark.xml").load(path);
        else if (fileFormat.equalsIgnoreCase(Constants.FIXED)){
          logger.info("Fixed");
             dataFrame = reader.text(path);
            //logger.info("Fixed");
            dataFrame = read_fixed_file(dataFrame,fileInput);}
        else if(fileFormat.equalsIgnoreCase(Constants.AVRO))
            dataFrame = reader.format(Constants.AVRO).load(path);
         else
           try {
               throw new ResourceNotFoundException(" Unknown data format! ");
           } catch (ResourceNotFoundException e) {
               e.printStackTrace();
           }

       logger.info("After Execution of Read File " + dateTimeUtilities.getCurrentDateTime());
       dataSetMap.put(dataFrameName,dataFrame);

       logger.info(" MapModel Read File  ...." + datasetModel.toString());

       return dataSetMap;
    }

    public Dataset<Row> read_fixed_file(Dataset<Row> dataFrame,FileInput fileInput){
        List<ColumnNameTypeLength> columnNameTypeLengthList = fileInput.getDataFormat().getColNameTypeLength();

        //StructField[] myArr = new StructField[];
        if(columnNameTypeLengthList != null && !columnNameTypeLengthList.isEmpty()){

             StructType schema = new StructType(columnNameTypeLengthList.stream().map(column -> {
                 return new StructField(column.getColumnName(), inferType(column.getColumnType()), true, Metadata.empty());
            }).toArray(StructField[]::new));
            System.out.println("before dataframe.show()");

            System.out.println("Schema for final df : " + schema.length());
            System.out.println("Schema for final df : " + schema.fields());

            JavaRDD parsedRdd = dataFrame.toJavaRDD().map(x -> {
                String str = x.getString(0);
                Integer startIndex = 0;
                List<Object> columnsDataList = new ArrayList<>();
                for (ColumnNameTypeLength column : columnNameTypeLengthList) {
                    Integer currPos = column.getColumnLength();
                    String columnData = "";
                    columnData = str.substring(startIndex, startIndex + currPos);
                    startIndex += currPos;
                    columnsDataList.add(getTypedData(columnData.trim(), column.getColumnType()));
                }
                Row row = RowFactory.create(columnsDataList.toArray());
                return row;
            });
            Dataset<Row> newdf = sparkSession.createDataFrame(parsedRdd,schema);



            newdf.printSchema();

            newdf.show();
            return newdf;

        };

        return newdf;
    }

    public Object getTypedData(String columnData, String type) {
        switch (type) {

            case "int":
                return Integer.parseInt(columnData);
            case "long":
                return Long.parseLong(columnData);
            case "str":
                return columnData;
            case "float":
                return Float.parseFloat(columnData);
            case "double":
                return Double.parseDouble(columnData);
            case "boolean":
                return Boolean.parseBoolean(columnData);
            case "date":
                return java.sql.Date.valueOf(columnData);
            default:
                return columnData;

        }


    }


    public DataType inferType(String columnType) {
        switch (columnType) {

            case "int":
                return DataTypes.IntegerType;
            case "long":
                return DataTypes.LongType;
            case "str":
                return DataTypes.StringType;
            case "float":
                return DataTypes.FloatType;
            case "double":
                return DataTypes.DoubleType;
            case "boolean":
                return DataTypes.BooleanType;
            case "date":
                return DataTypes.DateType;
            default:
                return DataTypes.StringType;
        }
    }
}
